File: frontend/src/utils/indicators.js
// compute values for SMA series
export function computeSMA(data, period) {
  const sma = [];
  for (let i = 0; i < data.length; i++) {
    if (i < period - 1) {
      sma.push({ time: data[i].time, value: null });
    } else {
      const slice = data.slice(i - period + 1, i + 1);
      const sum = slice.reduce((acc, d) => acc + d.close, 0);
      sma.push({ time: data[i].time, value: sum / period });
    }
  }
  return sma;
}

// compute values for EMA series
export function computeEMA(prices, period) {
  if (!prices || prices.length < period) return [];

  const k = 2 / (period + 1); // smoothing factor
  const emaArray = [];

  // Seed EMA with first SMA
  let sum = 0;
  for (let i = 0; i < period; i++) sum += prices[i].close;
  let emaPrev = sum / period;
  emaArray[period - 1] = { time: prices[period - 1].time, value: emaPrev };

  // Compute EMA for the rest
  for (let i = period; i < prices.length; i++) {
    const price = prices[i].close;
    emaPrev = price * k + emaPrev * (1 - k);
    emaArray[i] = { time: prices[i].time, value: emaPrev };
  }

  // Fill initial values with null for alignment
  for (let i = 0; i < period - 1; i++) {
    emaArray[i] = { time: prices[i].time, value: null };
  }

  return emaArray;
}

// compute values for bollinger bands
export function computeBollingerBands(prices, period = 20, stdDevMultiplier = 2) {
  if (!prices || prices.length < period) return [];

  const bands = [];

  for (let i = 0; i < prices.length; i++) {
    if (i < period - 1) {
      bands.push({ time: prices[i].time, upper: null, middle: null, lower: null });
      continue;
    }

    // Extract the window of values for this period
    const window = prices.slice(i - period + 1, i + 1).map(p => p.close);

    // Compute mean (middle band)
    const mean = window.reduce((a, b) => a + b, 0) / period;

    // Compute standard deviation
    const variance = window.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / period;
    const stdDev = Math.sqrt(variance);

    // Upper and lower bands
    const upper = mean + stdDevMultiplier * stdDev;
    const lower = mean - stdDevMultiplier * stdDev;

    bands.push({ time: prices[i].time, upper, middle: mean, lower });
  }

  return bands;
}

export function computeRSI(data, period = 14) {
  if (data.length < period) return [];

  let gains = [];
  let losses = [];
  let rsi = [];

  for (let i = 1; i < data.length; i++) {
    const change = data[i].close - data[i - 1].close;
    gains.push(change > 0 ? change : 0);
    losses.push(change < 0 ? -change : 0);

    if (i >= period) {
      const avgGain = gains.slice(-period).reduce((a, b) => a + b, 0) / period;
      const avgLoss = losses.slice(-period).reduce((a, b) => a + b, 0) / period;
      const rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
      const value = 100 - 100 / (1 + rs);

      rsi.push({ date: data[i].date, value });
    } else {
      rsi.push({ date: data[i].date, value: null });
    }
  }

  return rsi;
}

File: backend/app/utils/indicators.py
from math import sqrt
import pandas as pd

# ================================
# Simple Moving Average (SMA)
# ================================
def compute_sma_matrix(price_matrix: pd.DataFrame, period: int) -> pd.DataFrame:
    """
    Vectorized SMA for all symbols.
    """
    return price_matrix.rolling(window=period, min_periods=period).mean()


def compute_sma(data, period):
    """
    Computes the Simple Moving Average (SMA) for a given period.

    Parameters:
        data (list[dict]): List of dicts with keys 'date' and 'close'.
        period (int): Lookback period for the SMA calculation.

    Returns:
        list[dict]: Each element is a dict with:
            - 'date': corresponding date
            - 'value': SMA value (None for initial periods where SMA cannot be computed)
    """
    result = []
    closes = [d["close"] for d in data]
    dates = [d["date"] for d in data]

    for i in range(len(closes)):
        if i < period - 1:
            result.append({"date": dates[i], "value": None})  # Not enough data for SMA
        else:
            window = closes[i - period + 1:i + 1]
            sma_val = sum(window) / period
            result.append({"date": dates[i], "value": sma_val})
    return result


# ================================
# Bollinger Bands
# ================================
def compute_bollinger_bands_matrix(price_matrix: pd.DataFrame, period: int, std_dev: float):
    """
    Vectorized Bollinger Bands for all symbols.
    
    Returns a dict of DataFrames: {'upper', 'middle', 'lower'}
    """
    middle = compute_sma_matrix(price_matrix, period)
    rolling_std = price_matrix.rolling(window=period, min_periods=period).std()
    
    upper = middle + std_dev * rolling_std
    lower = middle - std_dev * rolling_std
    
    return {"upper": upper, "middle": middle, "lower": lower}


# ================================
# Exponential Moving Average (EMA)
# ================================
def compute_ema_matrix(price_matrix: pd.DataFrame, period: int) -> pd.DataFrame:
    """
    Vectorized EMA for all symbols.
    """
    return price_matrix.ewm(span=period, adjust=False).mean()


# ================================
# Relative Strength Index (RSI)
# ================================
def compute_rsi_matrix(price_matrix: pd.DataFrame, period: int) -> pd.DataFrame:
    """
    Vectorized RSI using Wilder's smoothing for all symbols.
    """
    delta = price_matrix.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)

    avg_gain = gain.ewm(alpha=1/period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1/period, adjust=False).mean()

    rs = avg_gain / avg_loss.replace(0, pd.NA)
    rsi = 100 - (100 / (1 + rs))
    return rsi.fillna(0)
